src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Waiting/VisualStudioOperationWaitIndicator.Instance.cs

```diff
@@ -0,0 +1,120 @@
+            {
+                (WaitIndicatorResult waitResult, _) = WaitForOperationImpl(title, message, allowCancel, token =>
+                {
+                    JoinableFactory.Run(() => asyncFunction(token));
+                    return true;
+                });
+
+                return waitResult;
+            }
+
+            public WaitIndicatorResult WaitForOperationWithResult(string title, string message, bool allowCancel, Action<CancellationToken> action)
+            {
+                (WaitIndicatorResult waitResult, _) = WaitForOperationImpl(title, message, allowCancel, token =>
+                {
+                    action(token);
+                    return true;
+                });
+
+                return waitResult;
+            }
+
+            public (WaitIndicatorResult, T) WaitForAsyncOperationWithResult<T>(string title, string message, bool allowCancel, Func<CancellationToken, Task<T>> asyncFunction)
+            {
+                return WaitForOperationImpl(title, message, allowCancel, token => JoinableFactory.Run(() => asyncFunction(token)));
+            }
+
+            public (WaitIndicatorResult, T) WaitForOperationWithResult<T>(string title, string message, bool allowCancel, Func<CancellationToken, T> function)
+            {
+                return WaitForOperationImpl(title, message, allowCancel, function);
+            }
+
+            private (WaitIndicatorResult, T) WaitForOperationImpl<T>(string title, string message, bool allowCancel, Func<CancellationToken, T> function)
+            {
+                _threadingService.VerifyOnUIThread();
+                using (IWaitContext waitContext = StartWait(title, message, allowCancel))
+                {
+                    try
+                    {
+                        T result = function(waitContext.CancellationToken);
+
+                        return (WaitIndicatorResult.Completed, result);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        return (WaitIndicatorResult.Canceled, default);
+                    }
+                    catch (AggregateException aggregate) when (aggregate.InnerExceptions.All(e => e is OperationCanceledException))
+                    {
+                        return (WaitIndicatorResult.Canceled, default);
+                    }
+                }
+            }
+
+            private IWaitContext StartWait(string title, string message, bool allowCancel)
+                => new VisualStudioWaitContext(_threadedWaitDialogFactory, title, message, allowCancel);
+
+            protected override Task DisposeCoreAsync(bool initialized) => Task.CompletedTask;
+            protected override Task InitializeCoreAsync(CancellationToken cancellationToken) => InitializeAsync();
+
+            public async Task InitializeAsync()
+            {
+                await _threadingService.SwitchToUIThread();
```

Comment:
Just call through IVsUIService.Value/GetValueAsync always - we already cache the result, you don't need to recache.